### ISSUE TITLE
sim-cse 里的 train dataset, 为啥要在 tokenize 阶段传 2 次 source ，而不在 training 时 对 input_ids 进行 repeat 呢？

### ISSUE THREAD

**Issue by ericxsun** [zhengyanzhao1997/NLP-model#2]
如题，

我简单实验了下， tokenize 时 传 2 次 source ，结果正常，loss 能快速下降，见代码 [tokenize source source](https://github.com/zhengyanzhao1997/NLP-model/blob/77d5f5d1ea4816cb674f939469914590927f4c48/model/model/Torch_model/SimCSE-Chinese/train_unsupervised.py#L68)
```
class TrainDataset(Dataset):
...
    def text_to_id(self, source):
        sample = self.tokenizer([source, source],max_length=self.maxlen,truncation=True,padding='max_length',return_tensors='pt')
        return sample
...
```

而 tokenize 时只传一次 source，且在 train 时 repeat input_ids, loss 一直不降
```
class TrainDataset(Dataset):
...
    def text_to_id(self, source):
        sample = self.tokenizer(source,max_length=self.maxlen,truncation=True,padding='max_length',return_tensors='pt')
        return sample
...

def train(dataloader,testdata, model, optimizer):
    ...
    for batch, data in enumerate(dataloader):
        input_ids = data['input_ids'].view(len(data['input_ids'])*2,-1).repeat(2,1).to(device)
        ...
```

**Comment by zhengyanzhao1997** [comment]
您好感谢你的提醒，如果在training时repeat 你需要重新设计你的label矩阵或者调整tokenids的位置。tokenize2次完全是为了方便哈。这label矩阵构建也比较清晰 两两一对的出现在batch里



发自我的iPhone


------------------ 原始邮件 ------------------
发件人: QinLuo ***@***.***&gt;
发送时间: 2021年5月14日 16:41
收件人: zhengyanzhao1997/NLP-model ***@***.***&gt;
抄送: Subscribed ***@***.***&gt;
主题: 回复：[zhengyanzhao1997/NLP-model] sim-cse 里的 train dataset, 为啥要在 tokenize 阶段传 2 次 source ，而不在 training 时 对 input_ids 进行 repeat 呢？ (#2)

**Comment by ericxsun** [comment]
> 您好感谢你的提醒，如果在training时repeat 你需要重新设计你的label矩阵或者调整tokenids的位置。tokenize2次完全是为了方便哈。这label矩阵构建也比较清晰 两两一对的出现在batch里 发自我的iPhone
> […](#)
> ------------------ 原始邮件 ------------------ 发件人: QinLuo ***@***.***&gt; 发送时间: 2021年5月14日 16:41 收件人: zhengyanzhao1997/NLP-model ***@***.***&gt; 抄送: Subscribed ***@***.***&gt; 主题: 回复：[zhengyanzhao1997/NLP-model] sim-cse 里的 train dataset, 为啥要在 tokenize 阶段传 2 次 source ，而不在 training 时 对 input_ids 进行 repeat 呢？ (#2)

嗯嗯，感谢。

但是看 [苏神的实现](https://github.com/bojone/SimCSE/blob/bce7175e9d87f45e6123d77d2080667ffa2915b4/eval.py#L116)直接是训练时在 data_generator 中 append  2 次，算 loss 的时候跟你也一样的，没有重新设计 label, 或 调整 token_ids 的位置，一样 loss 也下降很快。

**如果要重新设计 label / 调整 token_ids ，应该怎么调整呢？求指导，非常感谢。**

我尝试了在 train 时 repeat:

```
def repeat_self(ts):
    ts = torch.cat([torch.stack([t, t], 0) for t in ts], 0)
    return ts.to(device)

def train(dataloader,testdata, model, optimizer):
    ...
    for batch, data in enumerate(dataloader):
        input_ids = data['input_ids'].view(len(data['input_ids']),-1)
        input_ids = repeat_self(input_ids)
        ...
```

这样出来的数据，也和你的一样，2 个一组，类似：
```
101,5709,1446,6237,7370,2956,3326,749,2582,720,1215,102,0, ...., 0,0,0
101,5709,1446,6237,7370,2956,3326,749,2582,720,1215,102,0, ...., 0,0,0
101,955,1446,1282,753,4157,809,1400,2582,720,6820,3621,102,0, ...., 0,0,0
101,955,1446,1282,753,4157,809,1400,2582,720,6820,3621,102,0, ...., 0,0,0
101,6010,6009,955,1446,1962,719,2612,1908,7583,2428,102,0, ...., 0,0,0
101,6010,6009,955,1446,1962,719,2612,1908,7583,2428,102,0, ...., 0,0,0
...
```
但很奇怪，loss 也还是不下降

**Comment by zhengyanzhao1997** [comment]
如果是按照你这样的写法那就没问题了。不需要再作调整。可以检查下你的代码哈



发自我的iPhone


------------------ 原始邮件 ------------------
发件人: QinLuo ***@***.***&gt;
发送时间: 2021年5月14日 18:45
收件人: zhengyanzhao1997/NLP-model ***@***.***&gt;
抄送: zhengyanzhao1997 ***@***.***&gt;, Comment ***@***.***&gt;
主题: 回复：[zhengyanzhao1997/NLP-model] sim-cse 里的 train dataset, 为啥要在 tokenize 阶段传 2 次 source ，而不在 training 时 对 input_ids 进行 repeat 呢？ (#2)
